File: src/lib/notifications.ts
import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

export type NotificationKind =
  | "plan_created"
  | "plan_status_changed"
  | "idea_captured"
  | "idea_status_changed"
  | "category_created"
  | "schematic_updated"
  | "stage_started"
  | "stage_succeeded"
  | "stage_failed"
  | "stage_cancelled"
  | "run_started"
  | "run_finished"
  | "run_failed"
  | "integration_action"
  | "pending_question"
  | "schematic_drift_suspected";

export type NotificationDelivery = "toast_and_center" | "center_only" | "off";

export type Notification = {
  id: string;
  kind: NotificationKind;
  entityId: string;
  entityKind: string;
  projectPath: string;
  title: string;
  detail?: string;
  read: boolean;
  createdAt: number;
};

export type NotificationSettings = {
  overrides: Record<string, string>;
};

export async function notificationList(
  limit = 100,
  offset = 0,
): Promise<Notification[]> {
  return invoke<Notification[]>("notification_list", { limit, offset });
}

export async function notificationUnreadCount(): Promise<number> {
  return invoke<number>("notification_unread_count");
}

export async function notificationMarkRead(id: string): Promise<void> {
  await invoke("notification_mark_read", { id });
}

export async function notificationMarkAllRead(): Promise<void> {
  await invoke("notification_mark_all_read");
}

export async function notificationDelete(id: string): Promise<void> {
  await invoke("notification_delete", { id });
}

export async function notificationGetSettings(): Promise<NotificationSettings> {
  return invoke<NotificationSettings>("notification_get_settings");
}

export async function notificationSetSettings(
  settings: NotificationSettings,
): Promise<void> {
  await invoke("notification_set_settings", { settings });
}

/**
 * Subscribe to notification changes. The backend emits
 * `notifications://changed` after any mutation (insert, mark-read, delete).
 * The callback is invoked with no payload; callers should refetch.
 */
export function onNotificationsChanged(
  callback: () => void,
): Promise<UnlistenFn> {
  return listen("notifications://changed", () => callback());
}

File: src-tauri/src/commands/notifications.rs
use tauri::{AppHandle, Emitter};

use crate::{
    models::notification::{Notification, NotificationSettings},
    services::notification_service::NotificationService,
};

/// Emit a `notifications://changed` event so the frontend refreshes the
/// unread count + center after any mutation. Keep the payload minimal.
fn emit_changed(app: &AppHandle) {
    let _ = app.emit("notifications://changed", ());
}

#[tauri::command]
pub fn notification_list(
    limit: Option<i64>,
    offset: Option<i64>,
) -> Result<Vec<Notification>, String> {
    NotificationService::list(limit.unwrap_or(100), offset.unwrap_or(0))
}

#[tauri::command]
pub fn notification_unread_count() -> Result<i64, String> {
    NotificationService::unread_count()
}

#[tauri::command]
pub fn notification_mark_read(app: AppHandle, id: String) -> Result<(), String> {
    NotificationService::mark_read(&id)?;
    emit_changed(&app);
    Ok(())
}

#[tauri::command]
pub fn notification_mark_all_read(app: AppHandle) -> Result<(), String> {
    NotificationService::mark_all_read()?;
    emit_changed(&app);
    Ok(())
}

#[tauri::command]
pub fn notification_delete(app: AppHandle, id: String) -> Result<(), String> {
    NotificationService::delete(&id)?;
    emit_changed(&app);
    Ok(())
}

#[tauri::command]
pub fn notification_get_settings() -> Result<NotificationSettings, String> {
    NotificationService::get_settings()
}

#[tauri::command]
pub fn notification_set_settings(
    app: AppHandle,
    settings: NotificationSettings,
) -> Result<(), String> {
    NotificationService::set_settings(&settings)?;
    emit_changed(&app);
    Ok(())
}
